hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/SimpleConcurrentFileWritesConflictResolutionStrategy.java

```diff
@@ -18,65 +18,144 @@
+   *
+   * @param metaClient meta client
+   * @param currentInstant current instant
+   * @param pendingInstantsBeforeWrite pending instant recorded before write
+   * @return selected {@link HoodieInstant} stream
+   */
+  public Stream<HoodieInstant> getCompletedInstantsAfterCurrent(HoodieTableMetaClient metaClient, HoodieInstant currentInstant,
+                                                                Option<Set<String>> pendingInstantsBeforeWrite) {
+    return metaClient.getActiveTimeline()
+        .getTimelineOfActions(CollectionUtils.createSet(COMMIT_ACTION, REPLACE_COMMIT_ACTION, COMPACTION_ACTION, DELTA_COMMIT_ACTION))
         .filterCompletedInstants()
-        .findInstantsAfter(lastSuccessfulInstant.isPresent() ? lastSuccessfulInstant.get().getTimestamp() : HoodieTimeline.INIT_INSTANT_TS)
+        .filter(i -> compareTimestamps(i.getTimestamp(), GREATER_THAN, currentInstant.getTimestamp())
+            || (pendingInstantsBeforeWrite.isPresent() && pendingInstantsBeforeWrite.get().contains(i.getTimestamp())))
         .getInstantsAsStream();
+  }
+
+  /**
+   * Pick from instants that are still pending, need to compare the priority of the two operations.
+   *
+   * @param metaClient meta client
+   * @param currentInstant current instant
+   * @return selected {@link HoodieInstant} stream
+   */
+  public Stream<HoodieInstant> pickFromCurrentPendingInstants(HoodieTableMetaClient metaClient, HoodieInstant currentInstant) {
+
+    // Whether to pick the pending instant to candidate instants:
+    // +-----------------+----------------------------------------+---------------------------------------+------------+
+    // | current\pending | ingestion                              |  clustering                           | compaction |
+    // |-----------------+----------------------------------------+---------------------------------------+------------+
+    // | ingestion       | no                                     | no if #isIngestionPrimaryClustering() | yes        |
+    // |-----------------+----------------------------------------+---------------------------------------+------------+
+    // | clustering      | yes if #isIngestionPrimaryClustering() | no                                    | no         |
+    // +-----------------+----------------------------------------+---------------------------------------+------------+
+
+    Set<String> actionsToPick = new HashSet<>();
+
+    if (REPLACE_COMMIT_ACTION.equals(currentInstant.getAction())
+        && ClusteringUtils.isPendingClusteringInstant(metaClient, currentInstant)) {
+      if (isIngestionPrimaryClustering()) {
+        actionsToPick.add(COMMIT_ACTION);
+        actionsToPick.add(DELTA_COMMIT_ACTION);
+      }
+    } else {
+      if (!isIngestionPrimaryClustering()) {
+        actionsToPick.add(REPLACE_COMMIT_ACTION);
+      }
+      actionsToPick.add(COMPACTION_ACTION);
```

Comment:
Here I just keep the current implementation, because log compaction has pre commit now, hope to get more opinions @suryaprasanna